### DISCUSSION TITLE
rtc.datetime() : OSError: [Errno 22] EINVAL

### DISCUSSION THREAD

**Discussion by puppet13th** [micropython/micropython#10616] [RP2040 / Pico] [answered]
example code:
```
from machine import RTC
from time import localtime
rtc = RTC()

rtc.datetime( localtime() )
```

code output:
```
Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
OSError: [Errno 22] EINVAL
```

have tried using firmware 1.18 and 1.19.1. the code run fine on esp32 port.
boards : seeed xiao-rp2040 and waveshare rp2040-zero

**Comment by robert-hh** [comment, marked as answer]
time.localtime() and rtc.datetime() have a different format for the datetime tuple. See e.g. issues #10578 or #5553 and many more.
Or the documentation at https://docs.micropython.org/en/latest/library/time.html?highlight=time%20localtime#time.localtime and https://docs.micropython.org/en/latest/library/machine.RTC.html?highlight=datetime#machine.RTC.datetime.
rtc.datetime() is as well different from CPython's datetime.datetime() method.
Applying time.localtime() to rtc.datetime() may or may not cause an exception, depending on the  port's implementation, but in any case the result is wrong.

**Reply by puppet13th** [reply]
@robert-hh thank you for your reply.
So how can we set rtc date and time in rp2040 port?

In rp2040 port rtc only have datetime method.i assume it will set date and time if we provide date and time tuple like in esp32 port( just like in the documentation)

**Reply by robert-hh** [reply]
Just provide the datetime tuple as documented in the class method https://docs.micropython.org/en/latest/library/machine.RTC.html?highlight=rtc%20datetime#machine.RTC.datetime. The documentation at the quickref is wrong. See also issue #10578 and the PR to fix that. To set the date & time to Jan 29, 2023, 9:20.31,, use e.g.

```
from machine import RTC
rtc=RTC()
rtc.datetime((2023,1,29,0,9,20,31,0))
```
Note that the day-of-week item after the day is ignored and will be calculated from the date when calling rtc.datetime() to get the date & time.

**Reply by puppet13th** [reply]
@robert-hh thank you for your reply.
in esp32 port. rtc.datetime() just accept wrong datetime format from localtime and does not throw exception, that why i miss understood.

```py
from time import localtime, mktime
from machine import RTC

rtc = RTC()

print( localtime() )
print( rtc.datetime() )
rtc.datetime( localtime( mktime( localtime() ) ) )
print( localtime() )
print( rtc.datetime() )
```

code output (esp32):
```
(2023, 1, 29, 23, 11, 46, 6, 29)
(2023, 1, 29, 6, 23, 11, 46, 183418)
(2023, 1, 29, 11, 46, 6, 6, 29)
(2023, 1, 29, 6, 11, 46, 6, 317) --> wrong time
```

localtime tuple order need to be changed in order to be passed to rtc.datetime()

i'm just playing with seed xiao expansion board that have onboard PCF8563 rtc chip.
